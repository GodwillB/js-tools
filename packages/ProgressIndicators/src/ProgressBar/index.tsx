import React from 'react';

/** Props for ProgressBar */
interface ProgressBarProps {
  animate: boolean /** apply animation to progressBar */;
  decimalPoints: number /** Rounds off the value to this number */;
  height: string /** resizes the height of the progressBar */;
  min: number /** set lower bound for the progressBar range */;
  max: number /** set upper bound for the progressBar range */;
  progressStrippedClass: string /** set progressBar class to stripped */;
  progressAnimationClass: string /** set progressBar class to Animate */;
  value: number /** Represents the progress bar value */;
  lineColor: string /** set line colors */;
  lineColorThresholds?: { [key: string]: number } /** set linecolor threshold */;
  showLabel: boolean /** set label on progressBar */;
  stripped: boolean /** set strips in progressBar */;
}

/** default props for ProgressBar */
const defaultProgressBarProps = {
  animate: false,
  decimalPoints: 0,
  height: '10px',
  lineColor: '#0000FF',
  lineColorThresholds: undefined,
  max: 100,
  min: 0,
  progressAnimationClass: 'progress-bar-striped progress-bar-animated',
  progressStrippedClass: 'progress-bar-striped',
  showLabel: false,
  stripped: false,
  value: 0
};

/** Displays configurable progress bar
 * lineColor prop when you don't require different line colors
 * lineColorThresholds when you have to match lineColors to certain value thresholds.
 * LineColorThresholds will take precedence over lineColor if both are provided
 */
const ProgressBar = (props: ProgressBarProps) => {
  const {
    animate,
    decimalPoints,
    value,
    lineColor,
    stripped,
    lineColorThresholds,
    showLabel,
    progressStrippedClass,
    progressAnimationClass
  } = props;
  let backgroundColor = lineColor;
  const max = props.max || 100;
  const min = props.min || 0;
  let range = max - min;
  if (range <= 0) {
    range = 100;
  }
  const decimalValue = value / range;
  const percentValue = decimalValue * 100;
  const percentValueString = percentValue.toFixed(decimalPoints);

  // set the line color: if lineColorThresholds is not given; lineColor will be used
  if (lineColorThresholds) {
    // sort the color and their thresholds by the threshold value
    const AscendingThresholds = Object.entries(lineColorThresholds).sort((e1, e2) => e1[1] - e2[1]);
    // top to bottom check to see which color threshold is matched first by the percentValue
    AscendingThresholds.forEach(([key, threshold]) => {
      if (percentValue >= threshold) {
        backgroundColor = key;
      }
    });
  }

  return (
    <div className="progress">
      <div
        className={`progress-bar ${stripped ? progressStrippedClass : ''}
        ${animate ? progressAnimationClass : ''}`}
        style={{
          backgroundColor: `${backgroundColor}`,
          width: `${percentValueString}%`
        }}
        role="progressbar"
        aria-valuenow={percentValue}
        aria-valuemin={min}
        aria-valuemax={max}
      >
        {showLabel ? `${percentValue}%` : null}
      </div>
    </div>
  );
};

ProgressBar.defaultProps = defaultProgressBarProps;

export default ProgressBar;
