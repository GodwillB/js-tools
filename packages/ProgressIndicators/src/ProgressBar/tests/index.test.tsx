import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import ProgressBar from '../';
const RED = '#FF4136';
const YELLOW = '#FFFF00';
const BLUE = '#0000FF';
const GREY = '#dddddd';

describe('components/progressBar', () => {
  it('renders without crashing', () => {
    shallow(<ProgressBar />);
  });

  it('renders correctly for the default props', () => {
    const props = {
      value: 50
    };
    const wrapper = mount(<ProgressBar {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.progress-bar').props().style).toMatchSnapshot({
      backgroundColor: '#0000FF',
      width: '50%'
    });
    expect(wrapper.find('.progress-bar').props()).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly for this set of props', () => {
    const props = {
      decimalPoints: 0,
      height: '10px',
      lineColorThresholds: {
        GREEN_THRESHOLD: {
          color: '#2ECC40',
          name: 'Green',
          orEquals: true,
          value: 1
        },
        GREY_THRESHOLD: {
          color: '#dddddd',
          name: 'Grey',
          value: 0.2
        },
        RED_THRESHOLD: {
          color: '#FF4136',
          name: 'Red',
          orEquals: true,
          value: 0.75
        },
        YELLOW_THRESHOLD: {
          color: '#FFDC00',
          name: 'Yellow',
          value: 0.9
        }
      },
      max: 100,
      min: 0,
      value: 10
    };

    const wrapper = mount(<ProgressBar {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.progress-bar').props().style).toMatchSnapshot({
      backgroundColor: GREY,
      width: '10%'
    });
    wrapper.unmount();
  });

  it('renders correctly if both lineColor and threshold are given', () => {
    const props = {
      decimalPoints: 0,
      height: '10px',
      lineColor: '#5269EB',
      lineColorThresholds: {
        GREEN_THRESHOLD: {
          color: '#2ECC40',
          name: 'Green',
          orEquals: true,
          value: 1
        },
        GREY_THRESHOLD: {
          color: '#dddddd',
          name: 'Grey',
          value: 0.2
        },
        RED_THRESHOLD: {
          color: '#FF4136',
          name: 'Red',
          orEquals: true,
          value: 0.75
        },
        YELLOW_THRESHOLD: {
          color: '#FFDC00',
          name: 'Yellow',
          value: 0.9
        }
      },
      max: 100,
      min: 0,
      value: 40
    };

    const wrapper = mount(<ProgressBar {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.progress-bar').props().style).toMatchSnapshot({
      backgroundColor: RED,
      width: '40%'
    });
    wrapper.unmount();
  });

  it('renders correctly for different value', () => {
    const props = {
      decimalPoints: 0,
      height: '10px',
      lineColor: '#5269EB',
      lineColorThresholds: {
        GREEN_THRESHOLD: {
          color: '#2ECC40',
          name: 'Green',
          orEquals: true,
          value: 1
        },
        GREY_THRESHOLD: {
          color: '#dddddd',
          name: 'Grey',
          value: 0.2
        },
        RED_THRESHOLD: {
          color: '#FF4136',
          name: 'Red',
          orEquals: true,
          value: 0.75
        },
        YELLOW_THRESHOLD: {
          color: '#FFDC00',
          name: 'Yellow',
          value: 0.9
        }
      },
      max: 100,
      min: 0,
      value: 70
    };

    const wrapper = mount(<ProgressBar {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.progress-bar').props().style).toMatchSnapshot({
      backgroundColor: RED,
      width: '70%'
    });
    wrapper.unmount();
  });

  it('uses lineColor if thresholds are not given', () => {
    const props = {
      decimalPoints: 0,
      height: '10px',
      lineColor: YELLOW,
      max: 100,
      min: 0,
      value: 70
    };

    const wrapper = mount(<ProgressBar {...props} />);
    expect(wrapper.find('.progress-bar').props().style).toMatchSnapshot({
      backgroundColor: YELLOW,
      width: '70%'
    });
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('how does it behave when there is no value', () => {
    const props = {
      decimalPoints: 0,
      height: '10px'
    };
    const wrapper = mount(<ProgressBar {...props} />);
    expect(wrapper.find('.progress-bar').props().style).toMatchSnapshot({
      backgroundColor: BLUE,
      width: '0%'
    });
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('when the label value is provided', () => {
    const props = {
      showLabel: true,
      value: 60
    };
    const wrapper = mount(<ProgressBar {...props} />);
    expect(wrapper.text()).toEqual('60%');
    wrapper.unmount();
  });

  it('renders a progress bar with strips', () => {
    const props = {
      stripped: true
    };
    const wrapper = mount(<ProgressBar {...props} />);
    expect(wrapper.find('.progress-bar-striped')).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders an animated progress bar', () => {
    const props = {
      anime: true
    };
    const wrapper = mount(<ProgressBar {...props} />);
    expect(wrapper.find('.progress-bar-animated')).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
