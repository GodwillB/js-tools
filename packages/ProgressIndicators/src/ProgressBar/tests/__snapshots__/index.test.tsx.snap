// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/progressBar how does it behave when there is no value 1`] = `
Object {
  "backgroundColor": "#0000FF",
  "width": "0%",
}
`;

exports[`components/progressBar how does it behave when there is no value 2`] = `
<ProgressBar
  animate={false}
  decimalPoints={0}
  height="10px"
  lineColor="#0000FF"
  max={100}
  min={0}
  progressAnimationClass="progress-bar-striped progress-bar-animated"
  progressStrippedClass="progress-bar-striped"
  showLabel={false}
  stripped={false}
  value={0}
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="progress-bar 
        "
      role="progressbar"
      style={
        Object {
          "backgroundColor": "#0000FF",
          "width": "0%",
        }
      }
    />
  </div>
</ProgressBar>
`;

exports[`components/progressBar renders a progress bar with strips 1`] = `ReactWrapper {}`;

exports[`components/progressBar renders a progress bar with strips 2`] = `
<ProgressBar
  animate={false}
  decimalPoints={0}
  height="10px"
  lineColor="#0000FF"
  max={100}
  min={0}
  progressAnimationClass="progress-bar-striped progress-bar-animated"
  progressStrippedClass="progress-bar-striped"
  showLabel={false}
  stripped={true}
  value={0}
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="progress-bar progress-bar-striped
        "
      role="progressbar"
      style={
        Object {
          "backgroundColor": "#0000FF",
          "width": "0%",
        }
      }
    />
  </div>
</ProgressBar>
`;

exports[`components/progressBar renders an animated progress bar 1`] = `ReactWrapper {}`;

exports[`components/progressBar renders an animated progress bar 2`] = `
<ProgressBar
  animate={false}
  anime={true}
  decimalPoints={0}
  height="10px"
  lineColor="#0000FF"
  max={100}
  min={0}
  progressAnimationClass="progress-bar-striped progress-bar-animated"
  progressStrippedClass="progress-bar-striped"
  showLabel={false}
  stripped={false}
  value={0}
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="progress-bar 
        "
      role="progressbar"
      style={
        Object {
          "backgroundColor": "#0000FF",
          "width": "0%",
        }
      }
    />
  </div>
</ProgressBar>
`;

exports[`components/progressBar renders correctly for different value 1`] = `
<ProgressBar
  animate={false}
  decimalPoints={0}
  height="10px"
  lineColor="#5269EB"
  lineColorThresholds={
    Object {
      "#00FF00": 70,
      "#FF0000": 0,
      "#FFFF00": 30,
    }
  }
  max={100}
  min={0}
  progressAnimationClass="progress-bar-striped progress-bar-animated"
  progressStrippedClass="progress-bar-striped"
  showLabel={false}
  stripped={false}
  value={70}
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={70}
      className="progress-bar 
        "
      role="progressbar"
      style={
        Object {
          "backgroundColor": "#00FF00",
          "width": "70%",
        }
      }
    />
  </div>
</ProgressBar>
`;

exports[`components/progressBar renders correctly for different value 2`] = `
Object {
  "backgroundColor": "#00FF00",
  "width": "70%",
}
`;

exports[`components/progressBar renders correctly for the default props 1`] = `
<ProgressBar
  animate={false}
  decimalPoints={0}
  height="10px"
  lineColor="#0000FF"
  max={100}
  min={0}
  progressAnimationClass="progress-bar-striped progress-bar-animated"
  progressStrippedClass="progress-bar-striped"
  showLabel={false}
  stripped={false}
  value={50}
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={50}
      className="progress-bar 
        "
      role="progressbar"
      style={
        Object {
          "backgroundColor": "#0000FF",
          "width": "50%",
        }
      }
    />
  </div>
</ProgressBar>
`;

exports[`components/progressBar renders correctly for the default props 2`] = `
Object {
  "backgroundColor": "#0000FF",
  "width": "50%",
}
`;

exports[`components/progressBar renders correctly for the default props 3`] = `
Object {
  "aria-valuemax": 100,
  "aria-valuemin": 0,
  "aria-valuenow": 50,
  "children": null,
  "className": "progress-bar 
        ",
  "role": "progressbar",
  "style": Object {
    "backgroundColor": "#0000FF",
    "width": "50%",
  },
}
`;

exports[`components/progressBar renders correctly for this set of props 1`] = `
<ProgressBar
  animate={false}
  decimalPoints={0}
  height="10px"
  lineColor="#0000FF"
  lineColorThresholds={
    Object {
      "#00FF00": 70,
      "#FF0000": 0,
      "#FFFF00": 30,
    }
  }
  max={100}
  min={0}
  progressAnimationClass="progress-bar-striped progress-bar-animated"
  progressStrippedClass="progress-bar-striped"
  showLabel={false}
  stripped={false}
  value={10}
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={10}
      className="progress-bar 
        "
      role="progressbar"
      style={
        Object {
          "backgroundColor": "#FF0000",
          "width": "10%",
        }
      }
    />
  </div>
</ProgressBar>
`;

exports[`components/progressBar renders correctly for this set of props 2`] = `
Object {
  "backgroundColor": "#FF0000",
  "width": "10%",
}
`;

exports[`components/progressBar renders correctly if both lineColor and threshold are given 1`] = `
<ProgressBar
  animate={false}
  decimalPoints={0}
  height="10px"
  lineColor="#5269EB"
  lineColorThresholds={
    Object {
      "#00FF00": 70,
      "#FF0000": 0,
      "#FFFF00": 30,
    }
  }
  max={100}
  min={0}
  progressAnimationClass="progress-bar-striped progress-bar-animated"
  progressStrippedClass="progress-bar-striped"
  showLabel={false}
  stripped={false}
  value={40}
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={40}
      className="progress-bar 
        "
      role="progressbar"
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "width": "40%",
        }
      }
    />
  </div>
</ProgressBar>
`;

exports[`components/progressBar renders correctly if both lineColor and threshold are given 2`] = `
Object {
  "backgroundColor": "#FFFF00",
  "width": "40%",
}
`;

exports[`components/progressBar uses lineColor if thresholds are not given 1`] = `
Object {
  "backgroundColor": "#FFFF00",
  "width": "70%",
}
`;

exports[`components/progressBar uses lineColor if thresholds are not given 2`] = `
<ProgressBar
  animate={false}
  decimalPoints={0}
  height="10px"
  lineColor="#FFFF00"
  max={100}
  min={0}
  progressAnimationClass="progress-bar-striped progress-bar-animated"
  progressStrippedClass="progress-bar-striped"
  showLabel={false}
  stripped={false}
  value={70}
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={70}
      className="progress-bar 
        "
      role="progressbar"
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "width": "70%",
        }
      }
    />
  </div>
</ProgressBar>
`;
